src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,345 @@
+        http = credentials.authorize(httplib2.Http())
+        self.service = apiclient.discovery.build('drive', 'v3', http=http)
+
+        # get gdrive folder
+        path = storage_url[len('gdrive://'):].rstrip('/')
+        self.folder = self._lookup_file(path)
+        if self.folder is None or self.folder['mimeType'] != Backend.MIME_TYPE_FOLDER:
+            raise DanglingStorageURLError(self.path)
+
+    def _get_credentials(self):
+        credentials_filename = oauth2client.client._get_environment_variable_file()
+        if not credentials_filename:
+            credentials_filename = oauth2client.client._get_well_known_file()
+        if not credentials_filename:
+            return None
+        with open(credentials_filename) as f:
+            creds = json.load(f)
+
+        credentials = oauth2client.client.SignedJwtAssertionCredentials(
+            creds['client_email'],
+            creds['private_key'],
+            'https://www.googleapis.com/auth/drive',
+            sub=self.login)
+        return credentials
+
+    def _list_files(self, folder, fields):
+        # list folder children
+        page_token = None
+        while True:
+            results = self.service.files().list(
+                q="'{0}' in parents".format(folder['id']),
+                pageToken=page_token,
+                fields="nextPageToken, files(%s)" % fields).execute()
+            for f in results.get('files', []):
+                yield f
+
+            # get next page
+            page_token = results.get('nextPageToken', None)
+            if page_token is None:
+                break
+
+    def _lookup_file(self, name, folder=None):
+        if folder is None:
+            folder = { 'id': 'root', 'path': '/', 'mimeType': Backend.MIME_TYPE_FOLDER }
+
+        # find next path component
+        if len(name) == 0:
+            return folder
+        if name[0] == '/':
+            name = name[1:]
+        n = name.split('/', 1)
+        if len(n[0]) == 0:
+            return folder
+
+        # iterate over folder children
+        for f in self._list_files(folder, "id, name, mimeType, size, properties"):
+            f['path'] = folder['path'] + f['name']
+            if f['mimeType'] == Backend.MIME_TYPE_FOLDER:
+                f['path'] += '/'
+
+            #log.debug("{0} ({1}: {2})".format(f['path'], f['id'], f['mimeType']))
+            if f['name'] != n[0]:
+                continue
+
+            # found matching child
+            if len(n) == 1:
+                return f
+            elif f['mimeType'] == Backend.MIME_TYPE_FOLDER:
+                return self._lookup_file(n[1], f)
+            else:
+                return None
+
+        # nothing found
+        return None
+
+    def _create_file(self, body):
+        return self.service.files().create(body=body).execute()
+
+    def _copy_file(self, fileId, body):
+        return self.service.files().copy(fileId=fileId, body=body).execute()
+
+    def _delete_file(self, f):
```

Comment:
Why a separate function for this? And if the function is truly necessary, why does it take an `f` object instead of `f['id]`?